opnode/p2p/gossip.go

```diff
@@ -0,0 +1,427 @@
+		pubsub.WithPeerExchange(false),
+		pubsub.WithBlacklist(denyList),
+		pubsub.WithGossipSubParams(BuildGlobalGossipParams(cfg)),
+	)
+	// TODO: pubsub.WithPeerScoreInspect(inspect, InspectInterval) to update peerstore scores with gossip scores
+}
+
+func validationResultString(v pubsub.ValidationResult) string {
+	switch v {
+	case pubsub.ValidationAccept:
+		return "ACCEPT"
+	case pubsub.ValidationIgnore:
+		return "IGNORE"
+	case pubsub.ValidationReject:
+		return "REJECT"
+	default:
+		return fmt.Sprintf("UNKNOWN_%d", v)
+	}
+}
+
+func logValidationResult(self peer.ID, msg string, log log.Logger, fn pubsub.ValidatorEx) pubsub.ValidatorEx {
+	return func(ctx context.Context, id peer.ID, message *pubsub.Message) pubsub.ValidationResult {
+		res := fn(ctx, id, message)
+		var src interface{}
+		src = id
+		if id == self {
+			src = "self"
+		}
+		log.Debug(msg, "result", validationResultString(res), "from", src)
+		return res
+	}
+}
+
+type seenBlocks struct {
+	sync.Mutex
+	blockHashes []common.Hash
+}
+
+// hasSeen checks if the hash has been marked as seen, and how many have been seen.
+func (sb *seenBlocks) hasSeen(h common.Hash) (count int, hasSeen bool) {
+	sb.Lock()
+	defer sb.Unlock()
+	for _, prev := range sb.blockHashes {
+		if prev == h {
+			return len(sb.blockHashes), true
+		}
+	}
+	return len(sb.blockHashes), false
+}
+
+// markSeen marks the block hash as seen
+func (sb *seenBlocks) markSeen(h common.Hash) {
+	sb.Lock()
+	defer sb.Unlock()
+	sb.blockHashes = append(sb.blockHashes, h)
+}
+
+func BuildBlocksValidator(log log.Logger, cfg *rollup.Config) pubsub.ValidatorEx {
+
+	// Seen block hashes per block height
+	// uint64 -> *seenBlocks
+	blockHeightLRU, err := lru.New(100)
+	if err != nil {
+		panic(fmt.Errorf("failed to set up block height LRU cache: %v", err))
+	}
+
+	return func(ctx context.Context, id peer.ID, message *pubsub.Message) pubsub.ValidationResult {
+		// [REJECT] if the compression is not valid
+		outLen, err := snappy.DecodedLen(message.Data)
+		if err != nil {
+			log.Warn("invalid snappy compression length data", "err", err, "peer", id)
+			return pubsub.ValidationReject
+		}
+		if outLen > maxGossipSize {
+			log.Warn("possible snappy zip bomb, decoded length is too large", "decoded_length", outLen, "peer", id)
+			return pubsub.ValidationReject
+		}
+
+		res := msgBufPool.Get().(*[]byte)
+		defer msgBufPool.Put(res)
+		data, err := snappy.Decode((*res)[:0], message.Data)
+		if err != nil {
+			log.Warn("invalid snappy compression", "err", err, "peer", id)
+			return pubsub.ValidationReject
+		}
+		*res = data // if we ended up growing the slice capacity, fine, keep the larger one.
+
+		// message starts with compact-encoding secp256k1 encoded signature
+		signatureBytes, payloadBytes := data[:65], data[65:]
+
+		// [REJECT] if the block encoding is not valid
+		var payload l2.ExecutionPayload
+		if err := payload.UnmarshalSSZ(uint32(len(payloadBytes)), bytes.NewReader(payloadBytes)); err != nil {
```

Comment:
We do a `[]byte -> reader -> []byte -> copy to dest` decoding here because of the fact that we Marshall with readers/writers. It'd probably be more performant to do a reader based API on the decoding side.

Reply:
All the underlying `[]byte` are pooled and recycled, I don't think it's that bad with performance. Repeated `Read` calls to an interface may even be slower.